### PR TITLE
Use .NET SDK to generate deps.json for tools

### LINKED ISSUE #6087: DotNetCliToolReference can't target netcoreapp2.0 and have references to 1.x packages
## Steps to reproduce
Create a tool that targets `netcoreapp2.0` and has a dependency that eventually resolves to a `System.*` package from `netstandard1.x` TFM.  For example, retarget https://github.com/dotnet/cli/blob/master/TestAssets/TestPackages/dotnet-dependency-context-test/dotnet-dependency-context-test.csproj in our own test assets to `netcoreapp2.0`.

Reference the tool from a project and invoke it.

## Expected  behavior
The tool executes fine.

## Actual behavior
You get errors like the following:

```
Unhandled Exception: System.IO.FileLoadException: Could not load file or assembly 'System.Linq, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
   at Microsoft.DotNet.Cli.CommandLine.CommandLineApplication.Argument(String name, String description, Action`1 configuration, Boolean multipleValues)
   at Microsoft.DotNet.Cli.CommandLine.CommandLineApplication.Argument(String name, String description, Boolean multipleValues)
   at Microsoft.DotNet.Tools.DependencyInvoker.DotnetBaseParams.AddDotnetBaseParameters()
   at Microsoft.DotNet.Tools.DependencyInvoker.DotnetBaseParams..ctor(String name, String fullName, String description)
   at Microsoft.DotNet.Tools.DependencyInvoker.Program.Main(String[] args)
```

## Environment data
`dotnet --info` output:

```
.NET Command Line Tools (2.0.0-preview1-005473)

Product Information:
 Version:            2.0.0-preview1-005473
 Commit SHA-1 hash:  38f139fd8e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   F:\cli\artifacts\win10-x64\stage2\sdk\2.0.0-preview1-005473\
```

## Notes
This is because `Microsoft.NETCore.App` v2.0 now has all the `System.*` assemblies inside a single nuget package instead of spread out like they were in 1.x.  When we [generate the tool's deps.json](https://github.com/dotnet/cli/blob/70c65160f6163666dbe5f052b2331ec59582fce5/src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs#L326) file, we are not doing the conflict resolution that the happens when generating a normal project's deps.json.  Normally, there is a .targets file in NETStandard.Library that does the conflict resolution.  See https://github.com/dotnet/standard/blob/7c681d51a8c1a74a8467f32233e39e65b30bd3f4/Microsoft.Packaging.Tools/tasks/targets/Microsoft.Packaging.Tools.targets#L86-L94.

@livarcocc @dsplaisted @ericstj 

## Tasks

- [ ] Set TFM to use for tool restore in SDK (instead of using hardcoded NuGet version)
  - This will cause tools to roll forward to the CLI's bundled framework
- [ ] Escape hatch to allow restoring for an older TFM (ie to restore and run tool on 1.x)
  - Probably expressed as metadata on the `DotNetCliToolReference`
  - Probably needs to be designed with NuGet team
- [ ] Generate deps.json as part of restore instead of when tool is invoked
  - This will avoid duplication of logic to generate deps.json in CLI and SDK
  - This should also bring in the conflict resolution logic which will fix this bug

### PR DESCRIPTION
**Pull request by dsplaisted**
This PR calls into a project in the .NET SDK to create deps.json files for .NET CLI tools instead of doing so directly.  This reduces logic duplication and will fix #6087.

The corresponding SDK PR is https://github.com/dotnet/sdk/pull/1128.  Once both PRs have gone through review, we'll merge the SDK PR, and then update this CLI PR to also insert the updated SDK.  This PR doesn't yet include tests, but when it does they will fail in CI until the SDK is updated.

### CODE REVIEW

**Comment by dsplaisted** [comment]
@eerhardt @nguerrera @livarcocc I think this is ready for review now.  The corresponding SDK PR is https://github.com/dotnet/sdk/pull/1128, which I still have a bit of work (tests, code review feedback) to do on.

@natemcmaster FYI

File: TestAssets/TestPackages/dotnet-portable-v1/dotnet-portable-v1.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (livarcocc) [inline TestAssets/TestPackages/dotnet-portable-v1/dotnet-portable-v1.csproj]: We have a dotnet-back-compat.Tests test project that has tests for a 10 project running a 1.0 tool. The way we did, we can specify the TFM of the tool, so that it is easy to continue to add these tests. Would that tool replace this one here?

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/IPackagedCommandSpecFactory.cs [hunk 2]
```diff
@@ -8,22 +8,14 @@ namespace Microsoft.DotNet.Cli.Utils
...
-        CommandSpec CreateCommandSpecFromLibrary(
-            LockFileTargetLibrary toolLibrary,
-            string commandName,
-            IEnumerable<string> commandArguments,
-            IEnumerable<string> allowedExtensions,
-            string nugetPackagesRoot,
-            CommandResolutionStrategy commandResolutionStrategy,
-            string depsFilePath,
-            string runtimeConfigPath);
+        //  Code review TODO: Is it OK to make breaking changes to the CLI Utils API surface?
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/IPackagedCommandSpecFactory.cs]: Strictly speaking, no. There are people who may be using it.

But AFAIK, this is not used externally.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs [hunk 3]
```diff
@@ -21,9 +21,41 @@ private static void AddAditionalParameters(string commandPath, IList<string> arg
...
+                        string.Format(
+                            LocalizableStrings.IgnoringPreferCLIRuntimeFile,
+                            nameof(PackagedCommandSpecFactory),
+                            runtimeConfig.Framework.Version,
+                            muxer.SharedFxVersion));
+                }
+                else
+                {
+                    arguments.Add("--fx-version");
+                    arguments.Add(new Muxer().SharedFxVersion);
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs]: Use the muxer instance from above.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs [hunk 4]
```diff
@@ -323,11 +330,61 @@ private IEnumerable<string> GetPossiblePackageRoots(IProject project)
...
+
+            args.Add(toolDepsJsonGeneratorProject);
+            args.Add($"/p:ProjectAssetsFile=\"{toolLockFile.Path}\"");
+            args.Add($"/p:ToolName={toolLibrary.Name}");
+            args.Add($"/p:ProjectDepsFilePath={tempDepsFile}");
+
+            var toolTargetFramework = toolLockFile.Targets.First().TargetFramework.GetShortFolderName();
+            args.Add($"/p:TargetFramework={toolTargetFramework}");
+
+
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs]: NIT: extra space here.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs [hunk 5]
```diff
@@ -323,11 +330,61 @@ private IEnumerable<string> GetPossiblePackageRoots(IProject project)
...
+            msBuildExePath = string.IsNullOrEmpty(msBuildExePath) ?
+                Path.Combine(AppContext.BaseDirectory, "MSBuild.dll") :
+                msBuildExePath;
+
+            var result = new MSBuildForwardingAppWithoutLogging(args, msBuildExePath).Execute();
+
+            if (result != 0)
+            {
+                //  TODO: Can / should we show the MSBuild output if there is a failure?
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs]: As verbose maybe, I think that would be a useful debugging tool.

File: src/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    /// <summary>
+    /// A class which encapsulates logic needed to forward arguments from the current process to another process
+    /// invoked with the dotnet.exe host.
+    /// </summary>
+    internal class ForwardingAppImplementation
+    {
+        private const string s_hostExe = "dotnet";
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs]: For const, we just use something like HostExe.

File: src/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+        private readonly string _forwardApplicationPath;
+        private readonly IEnumerable<string> _argsToForward;
+        private readonly string _depsFile;
+        private readonly string _runtimeConfig;
+        private readonly string _additionalProbingPath;
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs]: Only one here?

File: src/dotnet/ForwardingApp.cs [hunk 8]
```diff
@@ -1,28 +1,14 @@
...
-using Microsoft.DotNet.Cli.Utils;
+using System.Text;
 
 namespace Microsoft.DotNet.Cli
 {
-    /// <summary>
-    /// A class which encapsulates logic needed to forward arguments from the current process to another process
-    /// invoked with the dotnet.exe host.
-    /// </summary>
     public class ForwardingApp
```
Comment (livarcocc) [inline src/dotnet/ForwardingApp.cs]: Why not just use the ForwaringAppImplementation? And call that one ForwardingApp?

If you are worried about the fact that this class is public, don't, we don't ship the CLI as a library.

File: src/dotnet/commands/dotnet-msbuild/MSBuildForwardingApp.cs [hunk 9]
```diff
@@ -10,39 +10,25 @@
 using Microsoft.DotNet.Cli;
 using Microsoft.DotNet.Cli.CommandLine;
 using System.Diagnostics;
+using Microsoft.DotNet.Cli.Utils;
 
 namespace Microsoft.DotNet.Tools.MSBuild
 {
     public class MSBuildForwardingApp
```
Comment (livarcocc) [inline src/dotnet/commands/dotnet-msbuild/MSBuildForwardingApp.cs]: Same comment as for ForwardingApp.

File: TestAssets/TestPackages/dotnet-portable/Program.cs [hunk 10]
```diff
@@ -10,6 +12,12 @@ public class Program
         public static void Main(string[] args)
         {
             Console.WriteLine("Hello Portable World!");
+
+            var coreAssembly = typeof(object).GetTypeInfo().Assembly;
+            string coreFolder = Path.GetDirectoryName(coreAssembly.Location);
+            string frameworkVersion = Path.GetFileName(coreFolder);
```
Comment (eerhardt) [inline TestAssets/TestPackages/dotnet-portable/Program.cs]: The more correct way to do this is to do it the way the Muxer does it:  [https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.Cli.Utils/Muxer.cs#L21-L22](https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.Cli.Utils/Muxer.cs#L21-L22)

use the `FX_DEPS_FILE` passed through the appcontext.  The reasoning is that servicing can make it so the framework assemblies are loaded from a "servicing" folder, and not from the shared fx folder.

**Comment by eerhardt** [comment]
> using Microsoft.DotNet.Tools.Common;

Copyright.  on all new files.

---
Refers to: src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs:1 in 6139036. [](commit_id = 6139036bd4eef8d33ba55d23d226e38d3ed7b62d, deletion_comment = False)

File: src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+using NuGet.Packaging;
+using NuGet.ProjectModel;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    static class LockFileExtensions
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs]: (nit) always declare visibility.  We follow the same coding guidelines as corefx:  [https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)  see #5

File: test/Microsoft.DotNet.Cli.Utils.Tests/GivenAProjectToolsCommandResolver.cs [hunk 12]
```diff
@@ -456,5 +457,12 @@ private ProjectToolsCommandResolver SetupProjectToolsCommandResolver()
 
             return projectToolsCommandResolver;
         }
+
+        private string GetToolDepsJsonGeneratorProject()
+        {
+            //  When using the product, the ToolDepsJsonGeneratorProject property is used to get this path, but for testing
+            //  we'll hard code the path inside the SDK since we don't have a project to evaluate here
+            return Path.Combine(new RepoDirectoriesProvider().Stage2Sdk, @"Sdks\Microsoft.NET.Sdk\build\GenerateDeps\GenerateDeps.proj");
```
Comment (eerhardt) [inline test/Microsoft.DotNet.Cli.Utils.Tests/GivenAProjectToolsCommandResolver.cs]: Can't use slashes across platforms.  Need to use `Path.Combine("one", "two", "three")`

File: test/dotnet.Tests/PackagedCommandTests.cs [hunk 13]
```diff
@@ -91,6 +91,82 @@ public void TestProjectToolIsAvailableThroughDriver(string appName, bool useCurr
...
+            testInstance = testInstance.WithRestoreFiles();
+
+            new BuildCommand()
+                .WithProjectDirectory(testInstance.Root)
+                .Execute()
+                .Should().Pass();
+
+            new GenericCommand(toolPrefersCLIRuntime ? "portable-v1-prefercli" : "portable-v1")
+                .WithWorkingDirectory(testInstance.Root)
+                .WithEnvironmentVariable("DOTNET_MULTILEVEL_LOOKUP", "0")
```
Comment (eerhardt) [inline test/dotnet.Tests/PackagedCommandTests.cs]: ?? Why is this necessary?  And why is it only necessary in these 2 tests?

File: test/dotnet.Tests/PackagedCommandTests.cs [hunk 14]
```diff
@@ -91,6 +91,82 @@ public void TestProjectToolIsAvailableThroughDriver(string appName, bool useCurr
...
+                .Should().Pass();
+
+            var result =
+                new DotnetCommand(DotnetUnderTest.WithBackwardsCompatibleRuntimes)
+                .WithWorkingDirectory(testInstance.Root)
+                .WithEnvironmentVariable("DOTNET_MULTILEVEL_LOOKUP", "0")
+                .Execute(toolPrefersCLIRuntime ? "portable-v1-prefercli" : "portable-v1");
+
+            result.Should().Pass()
+                .And.HaveStdOutContaining("I'm running on shared framework version 1.1.1!");
```
Comment (eerhardt) [inline test/dotnet.Tests/PackagedCommandTests.cs]: I guess I don't see how this makes sense.  If we are running on a CLI that is running "2.0", and the tool says "I prefer the CLI", why would it still run on `1.1.1`? Does "prefercliruntime" even make sense anymore?

File: src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        public static string GetPackageDirectory(this LockFile lockFile, LockFileTargetLibrary library)
+        {
+            var packageFolders = lockFile.GetNormalizedPackageFolders();
+
+            var packageFoldersCount = packageFolders.Count();
+            var userPackageFolder = packageFoldersCount == 1 ? string.Empty : packageFolders.First();
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs]: I don't think this is safe.

@emgarten - Can you weigh in?

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs [hunk 16]
```diff
@@ -21,9 +21,41 @@ private static void AddAditionalParameters(string commandPath, IList<string> arg
         {
             if(PrefersCliRuntime(commandPath))
             {
-                arguments.Add("--fx-version");
-                arguments.Add(new Muxer().SharedFxVersion);
+                var runtimeConfigFile = Path.ChangeExtension(commandPath, FileNameSuffixes.RuntimeConfigJson);
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs]: What happens if this file doesn't exist?

**Comment by eerhardt** [comment]
>             var dependencyContext = new DepsJsonBuilder()

Should be deleted?  Also the DepsJsonBuilder should be deleted as well.

---
Refers to: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs:329 in 6139036. [](commit_id = 6139036bd4eef8d33ba55d23d226e38d3ed7b62d, deletion_comment = False)

**Comment by eerhardt** [comment]
>             var dependencyContext = new DepsJsonBuilder()

And all supporting logic that is no longer necessary.

---
In reply to: [297493505](https://github.com/dotnet/cli/pull/6356#issuecomment-297493505) [](ancestors = 297493505)

---
Refers to: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs:329 in 6139036. [](commit_id = 6139036bd4eef8d33ba55d23d226e38d3ed7b62d, deletion_comment = False)

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs [hunk 17]
```diff
@@ -323,11 +330,61 @@ private IEnumerable<string> GetPossiblePackageRoots(IProject project)
...
-                dependencyContextWriter.Write(dependencyContext, fileStream);
+                var platformLibraryPath = toolLockFile.GetPackageDirectory(platformLibrary);
+
+                if (platformLibraryPath != null && buildRelativePath != null)
+                {
+                    //  Get rid of "_._" filename
+                    buildRelativePath = Path.GetDirectoryName(buildRelativePath);
+
+                    string platformLibraryBuildFolderPath = Path.Combine(platformLibraryPath, buildRelativePath);
+                    var platformLibraryPropsFile = Directory.GetFiles(platformLibraryBuildFolderPath, "*.props").FirstOrDefault();
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs]: There are a lot of assumptions about the internals of the `Microsoft.NETCore.App` package in this code.  Is it safe to make all these assumptions?

1. That the .props file we are interested in is in the same directory as the first "build" file.
2. That the .props file we are interested in is the first .props file in that folder.

/cc @ericstj

File: src/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs (hunk 7 above)
Comment (dsplaisted) [inline src/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs]: This is what `ForwardingApp` did previously, I just moved the logic around and wrapped it.

File: src/dotnet/ForwardingApp.cs (hunk 8 above)
Comment (dsplaisted) [inline src/dotnet/ForwardingApp.cs]: The issue was with the test projects that reference the dotnet project.  Originally I just had `ForwardingApp` be internal in Cli.Utils, but that meant that things that derived from it in the dotnet project had to be internal, which meant that the test projects for dotnet couldn't use those types anymore.  Rather than adding InternalsVisibleTo from the dotnet project to all the test projects, I decided to split it up into internal and public pieces as you see now.

File: TestAssets/TestPackages/dotnet-portable/Program.cs (hunk 10 above)
Comment (dsplaisted) [inline TestAssets/TestPackages/dotnet-portable/Program.cs]: I saw that logic, but I didn't know the reason behind it and the code seemed uglier.  Since there's a reason behind it, I'll update it to use `FX_DEPS_FILE`.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs (hunk 17 above)
Comment (ericstj) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs]: Depends on how temporary this is.  You should at least be able to use the package ID instead of `*` that's a safe assumption to make since it's a nuget requirement.

A slightly better pattern would be the following: `<packageFolder>\build\<tfm>\<packageId>.props`

A even better algo would be to use [FindBestItemGroup](https://github.com/NuGet/NuGet.Client/blob/1a1c474aab5c1e851b416c583ca4c6d727bd94e3/src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs#L85) passing in [ManagedCodeConventions.Patterns.MSBuildFiles](https://github.com/NuGet/NuGet.Client/blob/1a1c474aab5c1e851b416c583ca4c6d727bd94e3/src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs#L373), but not so sure you want to take a NuGet dependency here.

File: test/dotnet.Tests/PackagedCommandTests.cs (hunk 13 above)
Comment (dsplaisted) [inline test/dotnet.Tests/PackagedCommandTests.cs]: In 2.0, the dotnet host will look in your user folder and in Program files for shared frameworks, in addition to the path where dotnet.exe is.  These tests are supposed to test that tools targeting 1.x fail to run if the 1.x shared framework isn't installed, so this environment variable stops it from resolving to the shared frameworks that you might have installed globally on your machine.

Perhaps we should set this environment variable all the time for the CLI.  @livarcocc

File: test/dotnet.Tests/PackagedCommandTests.cs (hunk 14 above)
Comment (dsplaisted) [inline test/dotnet.Tests/PackagedCommandTests.cs]: CoreFx does not want to support running apps that target a previous major version of .NET Core on higher major versions.  So "prefercliruntime" will only apply across minor revisions.

File: src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs (hunk 15 above)
Comment (dsplaisted) [inline src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs]: It did seem fragile to me but this is the logic that was [already there](https://github.com/dotnet/cli/blob/c72ed9fcd2d36eeb3354d3dcc0f78ddb6599bbc5/src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactory.cs#L101-L106), I just moved it around.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs (hunk 17 above)
Comment (dsplaisted) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectToolsCommandResolver.cs]: The real solution is for NuGet to generate .props and .targets files for tools that it restores: https://github.com/NuGet/Home/issues/5037.  Then we can just import the files NuGet generates.

This logic is just to get something working until we get that feature from NuGet (though I'm not sure when that will be).

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs (hunk 16 above)
Comment (dsplaisted) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs]: Is it allowed for tools to not have a runtimeconfig file?  My thought at the time was that this was required, although thinking about it again it seems like it could be OK not to include this file if you have prefercliruntime.

If it's OK for a tool not to have a runtimeconfig file then we'd need another way to check what version of .NET Core it targets, in order to prevent rolling forward.  We might be able to look at the path in the NuGet package, or look at the TargetFrameworkAttribute on the assembly.

File: src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs (hunk 15 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs]: I chatted with @emgarten before putting it there.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs (hunk 16 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs]: All tools have it at this point and the guidance that we provide is for people to package it when they are packaging tools.

**Comment by livarcocc** [comment]
@MattGertz for approval. This is one of the PRs I talked to you about.

File: TestAssets/TestPackages/dotnet-portable/Program.cs [hunk 18]
```diff
@@ -10,6 +12,20 @@ public class Program
...
+
+            var depsFile = new FileInfo(GetDataFromAppDomain("FX_DEPS_FILE"));
+            string frameworkVersion = depsFile.Directory.Name;
+
+            Console.WriteLine($"I'm running on shared framework version {frameworkVersion}!");
+        }
+
+        public static string GetDataFromAppDomain(string propertyName)
+        {
+            var appDomainType = typeof(object).GetTypeInfo().Assembly?.GetType("System.AppDomain");
```
Comment (eerhardt) [inline TestAssets/TestPackages/dotnet-portable/Program.cs]: Nit: since this is only ever called on net Core, you can just use AppContext.GetData. This code is only necessary if you need to support full fx as well

File: src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs (hunk 15 above)
Comment (emgarten) [inline src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs]: I might be forgetting some context on why it was done this way, but what I would expect here is:

```csharp
FallbackPackagePathResolver resolver = null;

if (packageFoldersCount > 0)
{
  // use packageFolders from the assets file
  var userPackageFolder = packageFolders.FirstOrDefault();
  var fallbackPackageFolders = packageFolders.Skip(1);
  resolver  = new FallbackPackagePathResolver(userPackageFolder, fallbackPackageFolders);
}
else
{
  // assets file does not contain packageFolders, use default settings
  // alternatively this could throw with an error
  var pathContext = new NuGetPathContext(NullSettings.Instance);
  resolver = new FallbackPackagePathResolver(pathContext);
}



```

Comment (emgarten) [inline src/Microsoft.DotNet.Cli.Utils/Extensions/LockFileExtensions.cs]: If there was just one packages folder listed then it looks like the user packages folder would be skipped, possibly this works because the CLI always adds in the fallback folder now so this value is always two or higher.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs (hunk 16 above)
Comment (dsplaisted) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecFactoryWithCliRuntime.cs]: Yeah, I chatted about this with @eerhardt and it turns out that NuGet pack includes the runtimeconfig file by default.

**Comment by livarcocc** [comment]
@dotnet-bot Test this please

**Comment by dsplaisted** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by dsplaisted** [comment]
@dotnet-bot test Linux x64 Release

**Comment by dsplaisted** [comment]
@dotnet-bot test Ubuntu x64 Release Build

**Comment by livarcocc** [comment]
@dotnet-bot test Linux x64 Release

**Comment by livarcocc** [comment]
@dotnet-bot Test Linux x64 Release Build
